src/work/Work.h

```diff
@@ -21,67 +20,84 @@ namespace stellar
+    bool allChildrenWaiting() const;
+    bool anyChildRaiseFailure() const;
+    bool anyChildRunning() const;
+    bool hasChildren() const;
+
+  protected:
+    Work(Application& app, std::function<void()> callback, std::string name,
+         size_t retries = BasicWork::RETRY_A_FEW,
+         Execution order = WORK_PARALLEL);
+
     // Note: `shared_from_this` assumes there exists a shared_ptr to the
     // references class. This relates to who owns what in Work interface.
-    // Thus, `addWork` should be used to create work (then the parent holds
+    // Thus, `addWork` should be used to create Work (then the parent holds
     // the reference).
-
     template <typename T, typename... Args>
     std::shared_ptr<T>
     addWork(Args&&... args)
     {
-        // `wakeUp` is sufficient as a callback for any child Work of
-        // WorkScheduler
-        std::weak_ptr<Work> weak(
-            std::static_pointer_cast<Work>(shared_from_this()));
-        auto callback = [weak]() {
-            auto self = weak.lock();
-            if (self)
-            {
-                self->wakeUp();
-            }
-        };
-
-        auto child =
-            std::make_shared<T>(mApp, callback, std::forward<Args>(args)...);
+        auto child = std::make_shared<T>(mApp, wakeUpCallback(),
+                                         std::forward<Args>(args)...);
         addChild(child);
         wakeUp();
         return child;
     }
 
-    // Children status helper methods
-    // TODO (mlo) look into potentially using std::all_of/std::any_of
-    bool allChildrenSuccessful() const;
-    bool allChildrenDone() const;
-    bool anyChildRaiseFailure() const;
-    bool anyChildFatalFailure() const;
-    bool anyChildRunning() const;
-    bool hasChildren() const;
-
-  protected:
-    Work(Application& app, std::function<void()> callback, std::string name,
-         size_t retries = BasicWork::RETRY_A_FEW);
-
-    // TODO (mlo) needs to private eventually,
-    // see comment for WorkScheduler::~WorkScheduler
-    void clearChildren();
     State onRun() final;
     void onReset() final;
 
     // Implementers decide what they want to do: spawn more children,
-    // wait for all children to finish, or perform work
+    // wait for all children to finish, or perform Work
     virtual BasicWork::State doWork() = 0;
 
     // Provide additional cleanup logic for reset
     virtual void doReset();
+
+  private:
+    std::list<std::shared_ptr<BasicWork>> mChildren;
+    std::list<std::shared_ptr<BasicWork>>::const_iterator mNextChild;
+    Execution const mExecutionOrder;
+
+    std::shared_ptr<BasicWork> yieldNextRunningChild();
+    void addChild(std::shared_ptr<BasicWork> child);
+    void clearChildren();
+
+    friend class WorkSequence;
+};
+
+    /*
+     * WorkSequence is a helper class, that implementers can use if they
+     * wish to enforce the order of work execution.
+     */
+class WorkSequence : public Work
```

Comment:
It seems like the sequence pattern is used frequently in the code (specifically, don't proceed to next work until current work completed successfully), e.g. `PutSnapshotFilesWork`, `PutHistoryArchiveStateWork` or `CatchupWork`. It might be useful to have a mechanism like WorkSequence that implementers can use to enforce execution order. We have previously discussed with @MonsieurNicolas that one of the parameters for this class could be a vector of shared_ptrs to `BasicWork`, however upon creation of BasicWork objects, It wasn't possible to construct a proper notification callback, because WorkSequence instance did not exist at that point. That led to some weird manipulations with `mNotifyCallback`, which was intended to be private and only usable by `BasicWork`. Additionally, it required iteration over mChildren, which we already do in `Work::yieldNextRunningChild`. So instead, I've slightly modified `Work` class to allow for serial execution. With parallel execution, we iterate over children, then parent, whereas with serial execution we keep yielding same child while it's running, scheduling parent only after the child has reached a complete state, so that the parent can decide how to proceed. Some example of `WorkSequence` usage are `PutHistoryArchiveStateWork` and `GzipAndPutFilesWork`.